### PR TITLE
update for land IAU to ccpp-physics noahmp

### PR DESCRIPTION
**Pull request by tsga**
This ensures consistency between the NoahMP component model and the UFS WM/ccpp physics Noahmp after the land IAU capability is added to the later. Specifically the noahmpdrv_init function has three parameters (Land IAU related DDTs) passed as optional, enabling the component model to use the noahmpdrv_init function without any more changes right now.

Dependent on pull request: https://github.com/ufs-community/ccpp-physics/pull/222

### CODE REVIEW

**Comment by tsga** [comment]
Tagging relevant people here. Please add (tag) anyone you think should be aware.
@uturuncoglu, @barlage, @CoryMartin-NOAA, @ClaraDraper-NOAA, @CatherineThomas-NOAA, @junwang-noaa

**Review by CoryMartin-NOAA — APPROVED** [review]
I don't know this code well but these changes seem correct from what I can tell

File: drivers/ccpp/noahmpdrv.F90 [hunk 1]
```diff
@@ -100,9 +119,309 @@ subroutine noahmpdrv_init(lsm, lsm_noahmp, me, isot, ivegsrc, &
...
 
       end subroutine noahmpdrv_init
 
+!> \ingroup NoahMP_LSM
+!! \brief This subroutine is called before noahmpdrv_run 
+!!  to update states with iau increments, if available
+!! \section arg_table_noahmpdrv_timestep_init Argument Table
+!! \htmlinclude noahmpdrv_timestep_init.html
+!!
+subroutine noahmpdrv_timestep_init (itime, fhour, delt, km,  ncols,         &
```
Comment (uturuncoglu) [inline drivers/ccpp/noahmpdrv.F90:140]: @barlage @tsga I think this is a new routine. Does this need to be called by the component model? At this point, the component model is just calling init and run. What happens if we don't call from component. Maybe it is just called in the IAU case.

File: drivers/ccpp/noahmpdrv.meta [hunk 2]
```diff
@@ -96,6 +96,230 @@
   dimensions = ()
```
Comment (uturuncoglu) [inline drivers/ccpp/noahmpdrv.meta:96]: As we discussed both files need to be pushed to NoahMP EMC repository too.

**Review by uturuncoglu — COMMENTED** [review]
I have just put couple of comment to clarify the implementation.

File: drivers/ccpp/noahmpdrv.F90 (hunk 1 above)
Comment (barlage) [reply drivers/ccpp/noahmpdrv.F90:140]: I believe @tsga has this routine only doing IAU so it doesn't need to be called in the component model for now.

Comment (tsga) [reply drivers/ccpp/noahmpdrv.F90:140]: > I believe @tsga has this routine only doing IAU so it doesn't need to be called in the component model for now.

Yes, this function (and noahmpdrv_finalize) is only called for IAU, so the component model just skips it.

File: drivers/ccpp/noahmpdrv.meta (hunk 2 above)
Comment (tsga) [reply drivers/ccpp/noahmpdrv.meta:96]: > As we discussed both files need to be pushed to NoahMP EMC repository too.

Yes, they are now copied.

**Comment by barlage** [comment]
@tsga thanks for adding the IAU code here; I had a comment on that yesterday that apparently I didn't send; have you tried to compile and maybe run the land component reg test?

@uturuncoglu I assume there will need to be additions [here](https://github.com/NOAA-EMC/noahmp/blob/3ac32f0db7a2a97d930f44fa5f060c983ff31ee8/CMakeLists.txt#L21)

**Comment by tsga** [comment]
> @tsga thanks for adding the IAU code here; I had a comment on that yesterday that apparently I didn't send; have you tried to compile and maybe run the land component reg test?
> 
> @uturuncoglu I assume there will need to be additions [here](https://github.com/NOAA-EMC/noahmp/blob/3ac32f0db7a2a97d930f44fa5f060c983ff31ee8/CMakeLists.txt#L21)

@barlage yes, the land component related reg tests passed after all the changes were made. I am now running the whole UWM regression tests.  

And I have now added the CMakeLists.txt you mentioned to @uturuncoglu to noahmp repo as well (It is also part of the UWM repo, so I had modified it in the UWM repo--probably why the tests didn't fail).

**Comment by uturuncoglu** [comment]
@tsga Thanks. This looks fine. Just to be sure, did you run land component related RTs? If so, are they passing without any issue?

**Comment by tsga** [comment]
> @tsga Thanks. This looks fine. Just to be sure, did you run land component related RTs? If so, are they passing without any issue?

@uturuncoglu yes, All of the regression tests listed in "rt.conf" pass. Yesterday, "datm_cdeps_lm4_gswp3_intel" did fail, but it passes now after I updated the LM4-driver repo. The new land iau test (control_c48_lnd_iau_intel) needs to be run with the -c switch because the baseline outputs are not yet copied in hera (and the other test platforms).

**Comment by barlage** [comment]
@tsga what did you change in LM4? did you contact the LM4 group about those changes?

**Comment by tsga** [comment]
> @tsga what did you change in LM4? did you contact the LM4 group about those changes?

No, I didn't change anything. I only needed to get my branch up to date with all the develop branch of UFSWM and its submodules, and everything works OK.

**Comment by barlage** [comment]
> > @tsga what did you change in LM4? did you contact the LM4 group about those changes?
> 
> No, I didn't change anything. I only needed to get my branch up to date with all the develop branch of UFSWM and its submodules, and everything works OK.

OK, got it. I misinterpreted what you were saying.

**Review by ClaraDraper-NOAA — COMMENTED** [review]
Looks OK to me - just one small comment. Thanks @tsga

**Comment by ClaraDraper-NOAA** [comment]
@tsga I reviewed this repo as this is where the science is done. Would you like to review the related PRs too?

**Comment by tsga** [comment]
> @tsga I reviewed this repo as this is where the science is done. Would you like to review the related PRs too?

Thank you @ClaraDraper-NOAA. I think this is enough. The other two repo's will be reviewed by the code owners for dependency etc.

**Comment by tsga** [comment]
@uturuncoglu It seems the second test is failing because the files "lnd_iau_mod.F90" and "lnd_iau_mod.meta" are not copied to the noahmp directory. Can you please rerun the checks with the feature/lnd_iau branch contents and let me know if there is anything else to be done to merge this PR (the other related PRs are waiting for this to be complete).

**Comment by uturuncoglu** [comment]
@tsga Okay. Run it again. Let's see what happens.

File: CMakeLists.txt [hunk 3]
```diff
@@ -19,6 +19,7 @@ list(APPEND _noahmp_cap_files drivers/nuopc/lnd_comp_kind.F90
 
 # CCPP interface
 list(APPEND _noahmp_ccpp_files drivers/ccpp/noahmpdrv.F90
+	                       noahmp/drivers/ccpp/lnd_iau_mod.F90
```
Comment (uturuncoglu) [inline CMakeLists.txt]: @tsga I think this is wrong. It must be `drivers/ccpp/lnd_iau_mod.F90`. The test still fails.

Comment (tsga) [reply CMakeLists.txt]: Thank you @uturuncoglu. I just fixed it.

**Comment by tsga** [comment]
@barlage Mike, now that the tests are all passed, can you please mark this PR as approved.

**Review by barlage — APPROVED** [review]
Confirmed four files are identical to ccpp-physics

**Comment by FernandoAndrade-NOAA** [comment]
Testing for [2415](https://github.com/ufs-community/ufs-weather-model/pull/2415) has completed, please continue with the merge process, thank you.

**Comment by jkbk2004** [comment]
@uturuncoglu @barlage can you merge this pr?

**Comment by uturuncoglu** [comment]
@jkbk2004 merged.

**Pull request merged**